File: apps/moltnet-cli/wrapper_conformance_test.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"mvdan.cc/sh/v3/syntax"
)

// TestWrapperConformance asserts the shared cross-language fixtures: this Go
// guard and the TypeScript @themoltnet/shell-command-analyzer must resolve each
// command's prefix-runner (wrapper) chain to the same final executable. The
// fixture file is owned by the analyzer package and asserted by both, keeping
// the wrapper flag tables in sync without sharing parser code.
func TestWrapperConformance(t *testing.T) {
	t.Parallel()

	path := filepath.Join("..", "..", "libs", "shell-command-analyzer", "data", "wrapper-conformance.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture %s: %v", path, err)
	}
	var fixture struct {
		Cases []struct {
			Command string `json:"command"`
			Target  string `json:"target"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(data, &fixture); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	if len(fixture.Cases) == 0 {
		t.Fatal("no conformance cases found")
	}

	for _, c := range fixture.Cases {
		t.Run(c.Command, func(t *testing.T) {
			got := resolvePrefixRunnerTarget(t, c.Command)
			if got != c.Target {
				t.Fatalf("command %q: resolved executable %q, want %q", c.Command, got, c.Target)
			}
		})
	}
}

func resolvePrefixRunnerTarget(t *testing.T, command string) string {
	t.Helper()
	file, err := syntax.NewParser(syntax.Variant(syntax.LangBash)).Parse(strings.NewReader(command), "conformance")
	if err != nil {
		t.Fatalf("parse %q: %v", command, err)
	}
	var call *syntax.CallExpr
	syntax.Walk(file, func(node syntax.Node) bool {
		if call != nil {
			return false
		}
		if c, ok := node.(*syntax.CallExpr); ok && len(c.Args) > 0 {
			call = c
			return false
		}
		return true
	})
	if call == nil {
		t.Fatalf("no command found in %q", command)
	}
	executable, _, _, ok := parseShellInvocation(call, "")
	if !ok {
		t.Fatalf("could not resolve %q", command)
	}
	return filepath.Base(executable)
}

File: libs/shell-command-analyzer/__tests__/wrapper-conformance.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

import { beforeAll, describe, expect, it } from 'vitest';

import { analyzeCommand, initAnalyzer } from '../src/index.js';

/**
 * Shared cross-language conformance fixtures. The Go `github_guard`
 * (apps/moltnet-cli) asserts the same file — both must resolve each command's
 * prefix-runner chain to the same final executable, keeping the wrapper flag
 * tables in sync without sharing parser code. See the fixture's `description`.
 */
interface ConformanceFixture {
  cases: { command: string; target: string }[];
}

const fixture = JSON.parse(
  readFileSync(
    resolve(import.meta.dirname, '../data/wrapper-conformance.json'),
    'utf8',
  ),
) as ConformanceFixture;

beforeAll(async () => {
  await initAnalyzer();
});

describe('wrapper conformance (shared with apps/moltnet-cli github_guard)', () => {
  it('has fixtures', () => {
    expect(fixture.cases.length).toBeGreaterThan(0);
  });

  for (const { command, target } of fixture.cases) {
    it(`resolves ${JSON.stringify(command)} → ${target}`, async () => {
      const result = await analyzeCommand(command);
      if (!result.ok) {
        throw new Error(`expected ok for ${command}, got: ${result.reason}`);
      }
      // The final executable in a prefix-runner chain is the last resolved tool.
      const last = result.tools[result.tools.length - 1];
      expect(last.name).toBe(target);
    });
  }
});
